File: backend/app/services/community_service.py
"""
社区服务
"""

from typing import Optional, List
from app.db.supabase import get_supabase
from app.schemas.community import Community, CommunityCategory, CommunityListResponse, CommunityStats


class CommunityService:
    def __init__(self):
        self.db = get_supabase()

    def _format_community(
        self, data: dict, current_user_id: Optional[int] = None,
        real_member_count: Optional[int] = None,
        real_post_count: Optional[int] = None,
    ) -> Community:
        """格式化社区数据"""
        is_following = False
        if current_user_id:
            is_following = self._check_following(data["id"], current_user_id)

        return Community(
            id=data["id"],
            name=data["name"],
            slug=data["slug"],
            description=data.get("description", ""),
            iconUrl=data.get("icon_url", ""),
            coverUrl=data.get("cover_url", ""),
            category=data.get("category", "GENERAL"),
            isOfficial=data.get("is_official", False),
            isActive=data.get("is_active", True),
            memberCount=real_member_count if real_member_count is not None else data.get("member_count", 0),
            postCount=real_post_count if real_post_count is not None else data.get("post_count", 0),
            sortOrder=data.get("sort_order", 0),
            createdAt=data["created_at"],
            updatedAt=data["updated_at"],
            isFollowing=is_following,
        )

    def _check_following(self, community_id: int, user_id: int) -> bool:
        """检查用户是否关注了社区"""
        result = (
            self.db.table("community_follows")
            .select("id")
            .eq("community_id", community_id)
            .eq("user_id", user_id)
            .execute()
        )
        return bool(result.data)

    def _compute_real_counts(self, community_id: int) -> tuple:
        """动态计算社区的真实成员数和帖子数，并同步回缓存字段"""
        member_result = (
            self.db.table("community_follows")
            .select("id", count="exact")
            .eq("community_id", community_id)
            .execute()
        )
        real_member_count = member_result.count or 0

        post_result = (
            self.db.table("posts")
            .select("id", count="exact")
            .eq("community_id", community_id)
            .eq("status", "PUBLISHED")
            .eq("audit_status", "APPROVED")
            .execute()
        )
        real_post_count = post_result.count or 0

        try:
            self.db.table("communities").update({
                "member_count": real_member_count,
                "post_count": real_post_count,
            }).eq("id", community_id).execute()
        except Exception:
            pass

        return real_member_count, real_post_count

    def get_all_communities(
        self, current_user_id: Optional[int] = None
    ) -> List[Community]:
        """获取所有激活的社区"""
        result = (
            self.db.table("communities")
            .select("*")
            .eq("is_active", True)
            .order("sort_order", desc=True)
            .order("member_count", desc=True)
            .execute()
        )
        return [self._format_community(c, current_user_id) for c in result.data or []]

    def get_popular_communities(
        self, limit: int = 5, current_user_id: Optional[int] = None
    ) -> List[Community]:
        """获取热门社区（按成员数和帖子数综合排序）"""
        result = (
            self.db.table("communities")
            .select("*")
            .eq("is_active", True)
            .order("member_count", desc=True)
            .order("post_count", desc=True)
            .limit(limit)
            .execute()
        )
        return [self._format_community(c, current_user_id) for c in result.data or []]

    def get_following_communities(
        self, user_id: int, current_user_id: Optional[int] = None
    ) -> List[Community]:
        """获取用户关注的社区"""
        result = (
            self.db.table("community_follows")
            .select("community_id, communities(*)")
            .eq("user_id", user_id)
            .order("created_at", desc=True)
            .execute()
        )

        communities = []
        for item in result.data or []:
            c = item.get("communities")
            if c and c.get("is_active", True):
                communities.append(self._format_community(c, current_user_id))
        return communities

    def get_community_list(
        self, user_id: Optional[int] = None
    ) -> CommunityListResponse:
        """获取社区列表（热门、关注、全部）"""
        popular = self.get_popular_communities(limit=5, current_user_id=user_id)
        following = []
        if user_id:
            following = self.get_following_communities(user_id, user_id)
        all_communities = self.get_all_communities(user_id)

        return CommunityListResponse(
            popular=popular,
            following=following,
            all=all_communities,
        )

    def get_community_by_id(
        self, community_id: int, current_user_id: Optional[int] = None
    ) -> Optional[Community]:
        """获取单个社区（动态计算真实成员数和帖子数）"""
        result = (
            self.db.table("communities")
            .select("*")
            .eq("id", community_id)
            .execute()
        )
        if not result.data:
            return None

        real_member_count, real_post_count = self._compute_real_counts(community_id)
        return self._format_community(
            result.data[0], current_user_id,
            real_member_count=real_member_count,
            real_post_count=real_post_count,
        )

    def get_community_by_slug(
        self, slug: str, current_user_id: Optional[int] = None
    ) -> Optional[Community]:
        """通过 slug 获取社区（动态计算真实成员数和帖子数）"""
        result = (
            self.db.table("communities")
            .select("*")
            .eq("slug", slug)
            .execute()
        )
        if not result.data:
            return None

        community_id = result.data[0]["id"]
        real_member_count, real_post_count = self._compute_real_counts(community_id)
        return self._format_community(
            result.data[0], current_user_id,
            real_member_count=real_member_count,
            real_post_count=real_post_count,
        )

    def create_community(
        self,
        name: str,
        slug: str,
        description: str = "",
        icon_url: str = "",
        cover_url: str = "",
        category: str = "GENERAL",
        is_official: bool = False,
        sort_order: int = 0,
    ) -> Optional[Community]:
        """创建社区（管理员）"""
        insert_data = {
            "name": name,
            "slug": slug,
            "description": description,
            "icon_url": icon_url,
            "cover_url": cover_url,
            "category": category,
            "is_official": is_official,
            "sort_order": sort_order,
        }

        result = self.db.table("communities").insert(insert_data).execute()
        if not result.data:
            return None
        return self._format_community(result.data[0])

    def update_community(
        self, community_id: int, **kwargs
    ) -> Optional[Community]:
        """更新社区（管理员）"""
        update_data = {}
        field_mapping = {
            "name": "name",
            "description": "description",
            "icon_url": "icon_url",
            "cover_url": "cover_url",
            "category": "category",
            "is_official": "is_official",
            "is_active": "is_active",
            "sort_order": "sort_order",
        }

        for key, db_field in field_mapping.items():
            if key in kwargs and kwargs[key] is not None:
                update_data[db_field] = kwargs[key]

        if not update_data:
            return self.get_community_by_id(community_id)

        self.db.table("communities").update(update_data).eq("id", community_id).execute()
        return self.get_community_by_id(community_id)

    def delete_community(self, community_id: int) -> bool:
        """删除社区（管理员）"""
        result = (
            self.db.table("communities")
            .delete()
            .eq("id", community_id)
            .execute()
        )
        return bool(result.data)

    def follow_community(self, community_id: int, user_id: int) -> bool:
        """关注社区（幂等操作）"""
        print(f"[DEBUG] follow_community called: community_id={community_id}, user_id={user_id}")
        
        # 先检查是否已关注
        if self._check_following(community_id, user_id):
            print(f"[DEBUG] 用户 {user_id} 已关注社区 {community_id}，返回成功")
            return True  # 已关注，直接返回成功（幂等）
        
        try:
            # 插入关注记录
            insert_result = self.db.table("community_follows").insert(
                {"community_id": community_id, "user_id": user_id}
            ).execute()
            print(f"[DEBUG] 插入关注记录结果: {insert_result.data}")
            
            # 更新成员数（忽略错误，不影响关注成功）
            try:
                self.db.rpc(
                    "increment_community_member_count", {"community_id_param": community_id}
                ).execute()
            except Exception as rpc_error:
                print(f"[WARN] 更新成员数失败（不影响关注）: {rpc_error}")
            
            return True
        except Exception as e:
            print(f"[ERROR] 关注社区失败: {e}")
            import traceback
            traceback.print_exc()
            return False

    def unfollow_community(self, community_id: int, user_id: int) -> bool:
        """取消关注社区（幂等操作）"""
        # 先检查是否已关注
        if not self._check_following(community_id, user_id):
            return True  # 未关注，直接返回成功（幂等）
        
        try:
            result = (
                self.db.table("community_follows")
                .delete()
                .eq("community_id", community_id)
                .eq("user_id", user_id)
                .execute()
            )
            if result.data:
                # 更新成员数
                self.db.rpc(
                    "decrement_community_member_count", {"community_id_param": community_id}
                ).execute()
            return True
        except Exception as e:
            print(f"取消关注社区失败: {e}")
            return False

    def get_community_stats(self, community_id: int) -> Optional[CommunityStats]:
        """获取社区统计信息"""
        community = self.get_community_by_id(community_id)
        if not community:
            return None

        # 获取今日帖子数
        from datetime import datetime, timedelta
        today = datetime.now().replace(hour=0, minute=0, second=0, microsecond=0)
        week_ago = today - timedelta(days=7)

        today_result = (
            self.db.table("posts")
            .select("id", count="exact")
            .eq("community_id", community_id)
            .eq("status", "PUBLISHED")
            .gte("created_at", today.isoformat())
            .execute()
        )

        week_result = (
            self.db.table("posts")
            .select("id", count="exact")
            .eq("community_id", community_id)
            .eq("status", "PUBLISHED")
            .gte("created_at", week_ago.isoformat())
            .execute()
        )

        return CommunityStats(
            memberCount=community.memberCount,
            postCount=community.postCount,
            todayPostCount=today_result.count or 0,
            weekPostCount=week_result.count or 0,
        )

    def search_communities(
        self, keyword: str, limit: int = 20, current_user_id: Optional[int] = None
    ) -> List[Community]:
        """搜索社区"""
        result = (
            self.db.table("communities")
            .select("*")
            .eq("is_active", True)
            .or_(f"name.ilike.*{keyword}*,description.ilike.*{keyword}*")
            .order("member_count", desc=True)
            .limit(limit)
            .execute()
        )
        return [self._format_community(c, current_user_id) for c in result.data or []]


# 单例
community_service = CommunityService()

File: src/services/communityService.ts
/**
 * 社区服务
 */
import { config } from "../config/env";
import { useAuthStore } from "../store/authStore";

const EXPO_PUBLIC_API_BASE_URL = config.EXPO_PUBLIC_API_BASE_URL;

// API 响应包装类型
interface ApiResponse<T> {
  code: number;
  message: string;
  data: T;
}

// 社区分类
export type CommunityCategory = "GENERAL" | "FASHION" | "LIFESTYLE" | "BEAUTY" | "CULTURE";

// 社区类型
export interface Community {
  id: number;
  name: string;
  slug: string;
  description: string;
  iconUrl: string;
  coverUrl: string;
  category: CommunityCategory;
  isOfficial: boolean;
  isActive: boolean;
  memberCount: number;
  postCount: number;
  sortOrder: number;
  createdAt: string;
  updatedAt: string;
  isFollowing?: boolean;
}

// 社区列表响应
export interface CommunityListResponse {
  popular: Community[];
  following: Community[];
  all: Community[];
}

// 社区统计
export interface CommunityStats {
  memberCount: number;
  postCount: number;
  todayPostCount: number;
  weekPostCount: number;
}

// 通用请求方法 - 默认携带 token，支持自动刷新
async function request<T>(
  endpoint: string,
  options: RequestInit = {},
  isRetry: boolean = false
): Promise<T> {
  const url = `${EXPO_PUBLIC_API_BASE_URL}${endpoint}`;

  const headers: Record<string, string> = {
    "Content-Type": "application/json",
    Accept: "*/*",
    ...((options.headers as Record<string, string>) || {}),
  };

  // 自动添加 Authorization header（如果已登录）
  const authStore = useAuthStore.getState();
  let token = authStore.getAccessToken();

  // 如果 token 即将过期，先刷新
  if (
    token &&
    authStore.isTokenExpiringSoon &&
    authStore.isTokenExpiringSoon()
  ) {
    await authStore.refreshTokens();
    token = authStore.getAccessToken();
  }

  if (token) {
    headers["Authorization"] = `Bearer ${token}`;
  }

  const requestConfig: RequestInit = {
    ...options,
    headers,
  };

  try {
    const response = await fetch(url, requestConfig);
    const contentType = response.headers.get("content-type");

    if (!response.ok) {
      // 如果是 401 错误且不是重试请求，尝试刷新 token 后重试
      if (
        response.status === 401 &&
        !isRetry &&
        authStore.tokens?.refreshToken
      ) {
        const refreshSuccess = await authStore.refreshTokens();
        if (refreshSuccess) {
          return request<T>(endpoint, options, true);
        }
      }

      let errorMessage = "请求失败";

      if (contentType?.includes("application/json")) {
        const errorData = await response.json();
        errorMessage =
          errorData.detail ||
          errorData.message ||
          errorData.error ||
          errorMessage;
      } else {
        const text = await response.text();
        errorMessage = text || `HTTP ${response.status}`;
      }

      throw new Error(errorMessage);
    }

    if (contentType?.includes("application/json")) {
      const jsonResponse = await response.json();

      // 处理包装的 API 响应格式 { code, message, data }
      if (
        jsonResponse &&
        typeof jsonResponse === "object" &&
        "code" in jsonResponse
      ) {
        const apiResponse = jsonResponse as ApiResponse<T>;

        if (apiResponse.code !== 0) {
          throw new Error(apiResponse.message || "请求失败");
        }

        if ("data" in apiResponse) {
          return apiResponse.data;
        }
      }

      return jsonResponse as T;
    }

    const text = await response.text();
    return text as unknown as T;
  } catch (error) {
    if (error instanceof Error) {
      throw error;
    }
    throw new Error("网络请求失败，请检查网络连接");
  }
}

/**
 * 获取社区列表（热门、关注、全部）
 */
export async function getCommunities(): Promise<CommunityListResponse> {
  return request<CommunityListResponse>("/api/communities", {
    method: "GET",
  });
}

/**
 * 获取热门社区
 */
export async function getPopularCommunities(limit: number = 5): Promise<Community[]> {
  return request<Community[]>(`/api/communities/popular?limit=${limit}`, {
    method: "GET",
  });
}

/**
 * 获取我关注的社区
 */
export async function getFollowingCommunities(): Promise<Community[]> {
  return request<Community[]>("/api/communities/following", {
    method: "GET",
  });
}

/**
 * 获取社区详情
 */
export async function getCommunityById(communityId: number): Promise<Community> {
  return request<Community>(`/api/communities/${communityId}`, {
    method: "GET",
  });
}

/**
 * 通过 slug 获取社区详情
 */
export async function getCommunityBySlug(slug: string): Promise<Community> {
  return request<Community>(`/api/communities/slug/${slug}`, {
    method: "GET",
  });
}

/**
 * 获取社区统计信息
 */
export async function getCommunityStats(communityId: number): Promise<CommunityStats> {
  return request<CommunityStats>(`/api/communities/${communityId}/stats`, {
    method: "GET",
  });
}

/**
 * 关注社区
 */
export async function followCommunity(communityId: number): Promise<void> {
  return request<void>(`/api/communities/${communityId}/follow`, {
    method: "POST",
  });
}

/**
 * 取消关注社区
 */
export async function unfollowCommunity(communityId: number): Promise<void> {
  return request<void>(`/api/communities/${communityId}/follow`, {
    method: "DELETE",
  });
}

/**
 * 搜索社区
 */
export async function searchCommunities(keyword: string, limit: number = 20): Promise<Community[]> {
  return request<Community[]>(
    `/api/communities/search?keyword=${encodeURIComponent(keyword)}&limit=${limit}`,
    {
      method: "GET",
    }
  );
}

// 导出 communityService 对象
export const communityService = {
  getCommunities,
  getPopularCommunities,
  getFollowingCommunities,
  getCommunityById,
  getCommunityBySlug,
  getCommunityStats,
  followCommunity,
  unfollowCommunity,
  searchCommunities,
};

export default communityService;
